cpp/src/groupby/sort/sort_helper.cu

```diff
@@ -0,0 +1,312 @@
+    // in the row bitmask. When `_ignore_null_keys == true`, then only rows `[0, n)` 
+    // in the sorted keys are considered for grouping. 
+    _num_keys = keys_bitmask_column(stream).size() - 
+                keys_bitmask_column(stream).null_count();
+  } else {
+    _num_keys = _keys.num_rows();
+  }
+
+  return _num_keys; 
+}
+
+column_view helper::key_sort_order(cudaStream_t stream) {
+  if (_key_sorted_order)
+    return _key_sorted_order->view();
+
+  // TODO (dm): optimization. When keys are pre sorted but ignore nulls is true,
+  //            we still want all rows with nulls in the end. Sort is costly, so
+  //            do a copy_if(counting, sorted_order, {bitmask.isvalid(i)})
+  if (_keys_pre_sorted) {
+    _key_sorted_order = make_numeric_column(data_type(type_to_id<size_type>()),
+                          _keys.num_rows(), mask_state::UNALLOCATED, stream);
+
+    auto d_key_sorted_order = _key_sorted_order->mutable_view().data<size_type>();
+
+    thrust::sequence(rmm::exec_policy(stream)->on(stream), 
+                     d_key_sorted_order,
+                     d_key_sorted_order + _key_sorted_order->size(), 0);
+
+    return _key_sorted_order->view();
+  }
+
+  if (not _ignore_null_keys || !cudf::has_nulls(_keys)) {  // SQL style
+    _key_sorted_order = cudf::experimental::detail::sorted_order(_keys, {},
+      std::vector<null_order>(_keys.num_columns(), null_order::AFTER),
+      rmm::mr::get_default_resource(), stream);
+  } else {  // Pandas style
+    // Temporarily prepend the keys table with a column that indicates the 
+    // presence of a null value within a row. This allows moving all rows that 
+    // contain a null value to the end of the sorted order. 
+
+    auto augmented_keys = table_view({ 
+      table_view( {keys_bitmask_column()} ),
+      _keys });
+
+    _key_sorted_order = cudf::experimental::detail::sorted_order(
+      augmented_keys, {},
+      std::vector<null_order>(_keys.num_columns() + 1, null_order::AFTER),
+      rmm::mr::get_default_resource(), stream);
+
+    // All rows with one or more null values are at the end of the resulting sorted order.
+  }
+
+  return _key_sorted_order->view();
+}
+
+helper::index_vector const& helper::group_offsets(cudaStream_t stream) {
+  if (_group_offsets)
+    return *_group_offsets;
+
+  _group_offsets = std::make_unique<index_vector>(num_keys(stream));
+
+  auto device_input_table = table_device_view::create(_keys, stream);
+  auto sorted_order = key_sort_order().data<size_type>();
+  decltype(_group_offsets->begin()) result_end;
+  auto exec = rmm::exec_policy(stream);
+
+  if (has_nulls(_keys)) {
+    result_end = thrust::unique_copy(exec->on(stream),
+      thrust::make_counting_iterator<size_type>(0),
+      thrust::make_counting_iterator<size_type>(num_keys(stream)),
+      _group_offsets->begin(),
+      permuted_row_equality_comparator<true>(*device_input_table, sorted_order));
```

Comment:
It's the reverse. Permutation iterator lets us do this: `count_it[sort_order]` but we want this: `sort_order[count_it]`

Reply:
And to further clarify, we cannot simply use `sort_order.begin<T>()` because `thrust::unique_copy` will copy the iterator into the new array and we want the `count_it` copied, not the `sort_order`

Reply:
So I just had my `make_permutation_iterator` reversed? Or is the permutation iterator not possible to use?

Reply:
It's not possible to use unfortunately. We want to iterate over `key_sorted_order` but we want to store their indices rather than their values themselves.